### PR TITLE
feat: [PL-23322]: PreferenceFramework to retain scoped preferences in localStorage

### PR DESCRIPTION
**Pull request by mayankVermaHarness**
##### Jira Links:

https://harness.atlassian.net/browse/PL-23322

##### Screenshots:

You can use the following comments to re-trigger PR Checks

- Jest: `retrigger jest`
- Prettier: `retrigger prettier`
- Type Check: `retrigger typecheck`
- ESLint: `retrigger eslint`
- Standards: `retrigger standards`
- Build: `retrigger build`
- Title Check: `retrigger titlecheck`
- Coverage: `retrigger coverage`
- Rebase: `trigger rebase`
- Cypress: `trigger cypress`
- Fix Prettier - `fix prettier`

### CODE REVIEW

**Review by avinashmadhwani02 — CHANGES_REQUESTED** [review]

File: src/modules/10-common/utils/testUtils.tsx [hunk 1]
```diff
@@ -141,83 +142,92 @@ export const TestWrapper: React.FC<TestWrapperProps> = props => {
...
-            FEATURE_ENFORCEMENT_ENABLED: true
-          },
-          updateAppStore: () => void 0,
-          currentUserInfo: {},
-          ...defaultAppStoreValues
-        }}
-      >
-        <LicenseStoreContext.Provider
+    <Router history={history}>
+      <StringsContext.Provider value={{ data: stringsData as any, getString }}>
```
Comment (diptiman-adak) [inline src/modules/10-common/utils/testUtils.tsx:146]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232308)
```suggestion
    <StringsContext.Provider value={{ data: stringsData as unknown, getString }}>
```

Comment (diptiman-adak) [inline src/modules/10-common/utils/testUtils.tsx:146]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232308)
```suggestion
    <StringsContext.Provider value={{ data: stringsData as unknown, getString }}>
```

File: src/framework/PreferenceStore/PreferenceStoreContext.tsx [hunk 2]
```diff
@@ -0,0 +1,185 @@
...
+  const setPreference = (key: string, value: unknown, options?: PreferenceStoreOptions): void => {
+    if (options?.fromBackend) {
+      // TODO: ENHANCEMENT: call backend to set
+    } else {
+      const newPreferences = { ...currentPreferences, [key]: value }
+      setPreferences(newPreferences)
+    }
+  }
+
+  const getPreference = (key: string, options?: PreferenceStoreOptions): any => {
```
Comment (diptiman-adak) [inline src/framework/PreferenceStore/PreferenceStoreContext.tsx:126]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232308)
```suggestion
  const getPreference = (key: string, options?: PreferenceStoreOptions): unknown => {
```

File: src/framework/PreferenceStore/PreferenceStoreContext.tsx [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+export interface PreferenceStoreContextProps<T> {
+  preference: T
+  setPreference: (value: T) => void
+  clearPreference: () => void
+  updatePreferenceStore: (data: PreferenceStoreStateProps) => void
+}
+
+export const PREFERENCES_TOP_LEVEL_KEY = 'preferences'
+
+export const PreferenceStoreContext = React.createContext<PreferenceStoreProps<any>>({
```
Comment (diptiman-adak) [inline src/framework/PreferenceStore/PreferenceStoreContext.tsx:61]: Codacy has a fix for the issue: [Unexpected any. Specify a different type.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232308)
```suggestion
export const PreferenceStoreContext = React.createContext<PreferenceStoreProps<unknown>>({
```

File: src/framework/PreferenceStore/__tests__/PreferenceStoreContext.test.tsx [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+  const {
+    preference: savedVal,
+    setPreference: setSavedVal,
+    clearPreference,
+    updatePreferenceStore
+  } = usePreferenceStore<string>(PreferenceScope.MACHINE, ENTITY_TO_SAVE)
+
+  useEffect(() => {
+    updatePreferenceStore({ currentUserInfo: { email: 'abc@email.com' } })
+  }, [])
```
Comment (diptiman-adak) [inline src/framework/PreferenceStore/__tests__/PreferenceStoreContext.test.tsx:30]: Codacy has a fix for the issue: [React Hook useEffect has a missing dependency: 'updatePreferenceStore'. Either include it or remove the dependency array.](https://app.codacy.com/gh/harness/harness-core-ui/pullRequest?prid=9232308)
```suggestion
  }, [updatePreferenceStore])
```

**Comment by mayankVermaHarness** [comment]
retrigger build

**Comment by mayankVermaHarness** [comment]
trigger rebase

**Comment by mayankVermaHarness** [comment]
trigger cypress

File: src/framework/AppStore/AppStoreContext.tsx [hunk 5]
```diff
@@ -64,28 +75,57 @@ export function useAppStore(): AppStoreContextProps {
...
+  } = useParams<ProjectPathProps>()
+  let projectIdentifier = projectIdentifierFromPath
+  let orgIdentifier = orgIdentifierFromPath
+
+  const {
+    preference: savedProject,
+    setPreference: setSavedProject,
+    clearPreference: clearSavedProject,
+    updatePreferenceStore
+  } = usePreferenceStore<SavedProjectDetails>(PreferenceScope.USER, 'savedProject')
```
Comment (avinashmadhwani02) [inline src/framework/AppStore/AppStoreContext.tsx:109]: `savedProject` can be moved to a constant file

**Review by avinashmadhwani02 — APPROVED** [review]

**Review by sunnykesh-harness — APPROVED** [review]

**Comment by abhinavrastogi-harness** [comment]
retrigger jest

**Pull request merged**